File: epi_judge_cpp/sorted_list_to_bst.cc
#include <iterator>
#include <memory>
#include <vector>
#include "doubly_list_node.h"
#include "test_framework/generic_test.h"
#include "test_framework/test_failure.h"
#include "test_framework/timed_executor.h"
using std::make_shared;
using std::vector;

// Returns the root of the corresponding BST. The prev and next fields of the
// list nodes are used as the BST nodes left and right fields, respectively.
// The length of the list is given.
shared_ptr<ListNode<int>> BuildBSTFromSortedDoublyList(
    shared_ptr<ListNode<int>> l, int length) {
  // TODO - you fill in here.
  return nullptr;
}
void CompareVectorAndTree(const shared_ptr<ListNode<int>>& tree,
                          vector<int>::const_iterator& current,
                          const vector<int>::const_iterator& end) {
  if (!tree) {
    return;
  }

  CompareVectorAndTree(tree->prev, current, end);

  if (current == end) {
    throw TestFailure("Too few values in the tree");
  }
  if (*current != tree->data) {
    throw TestFailure("Unexpected value");
  }
  ++current;

  CompareVectorAndTree(tree->next, current, end);
}

void BuildBSTFromSortedDoublyListWrapper(TimedExecutor& executor,
                                         const vector<int>& l) {
  shared_ptr<ListNode<int>> input_list;
  for (auto it = rbegin(l); it != rend(l); ++it) {
    input_list = make_shared<ListNode<int>>(*it, nullptr, input_list);
    if (input_list->next) {
      input_list->next->prev = input_list;
    }
  }

  input_list = executor.Run([&] {
    return BuildBSTFromSortedDoublyList(input_list, static_cast<int>(l.size()));
  });

  auto current = begin(l);
  CompareVectorAndTree(input_list, current, end(l));
  if (current != end(l)) {
    throw TestFailure("Too many values in the tree");
  }

  while (input_list) {
    input_list->prev.reset();
    input_list = input_list->next;
  }
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"executor", "l"};
  return GenericTestMain(
      args, "sorted_list_to_bst.cc", "sorted_list_to_bst.tsv",
      &BuildBSTFromSortedDoublyListWrapper, DefaultComparator{}, param_names);
}

File: epi_judge_python/sorted_list_to_bst.py
import functools
from typing import Optional

from doubly_list_node import DoublyListNode
from test_framework import generic_test
from test_framework.test_failure import TestFailure
from test_framework.test_utils import enable_executor_hook


# Returns the root of the corresponding BST. The prev and next fields of the
# list nodes are used as the BST nodes left and right fields, respectively.
# The length of the list is given.
def build_bst_from_sorted_doubly_list(l: DoublyListNode,
                                      n: int) -> Optional[DoublyListNode]:
    # TODO - you fill in here.
    return None


def compare_vector_and_tree(tree, it):
    if not tree:
        return

    compare_vector_and_tree(tree.prev, it)

    v = next(it, None)
    if v is None:
        raise TestFailure('Too few values in the tree')
    if v != tree.data:
        raise TestFailure('Unexpected value')

    compare_vector_and_tree(tree.next, it)


@enable_executor_hook
def build_bst_from_sorted_doubly_list_wrapper(executor, l):
    input_list = None
    for v in reversed(l):
        input_list = DoublyListNode(v, next=input_list)
        if input_list.next != None:
            input_list.next.prev = input_list

    input_list = executor.run(
        functools.partial(build_bst_from_sorted_doubly_list, input_list,
                          len(l)))

    it = iter(l)
    compare_vector_and_tree(input_list, it)
    if next(it, None) is not None:
        raise TestFailure('Too many l in the tree')


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main(
            'sorted_list_to_bst.py', 'sorted_list_to_bst.tsv',
            build_bst_from_sorted_doubly_list_wrapper))
